File: question/leetcode/leetcode0139.Word-Break/leetcode0139.ts
function wordBreak(s: string, wordDict: string[]): boolean {
	const wordSet = new Set(wordDict);
	const dp: boolean[] = new Array(s.length + 1).fill(false);
	dp[0] = true; // dp[i]前i个字符是否可以被字典中的单词完全切分出来

	for (let i = 1; i <= s.length; i++) {
		for (let j = 0; j < i; j++) {
			const word = s.slice(j, i);
			if (dp[j] && wordSet.has(word)) {
				dp[i] = true;
				break;
			}
		}
	}

	return dp[s.length];
};

File: question/leetcode/leetcode0139.Word-Break/leetcode0139.java
import java.util.List;
import java.util.Set;
import java.util.HashSet;

class Solution {
	public boolean wordBreak(String s, List<String> wordDict) {
		Set<String> wordSet = new HashSet<>(wordDict);
		boolean[] dp = new boolean[s.length() + 1];
		dp[0] = true;

		for (int i = 0; i <= s.length(); i++) {
			for (int j = 0; j < i; j++) {
				String word = s.substring(j, i);
				if (dp[j] && wordSet.contains(word)) {
					dp[i] = true;
					break;
				}

			}
		}
		return dp[s.length()];
	}
}